### ISSUE TITLE
EPIC: Api poc

### ISSUE THREAD

**Issue by mikeAdamss** [GSS-Cogs/idpd-api-poc#63]
# API Poc: Initial

A fairly rough order of things to do to build this for late November.

- [x] Implement api with stubbed metadata store.
- [x] Implement api with stubbed csv store.

### Schemas

_Each jsonld has (or will have) an [associated schema](https://github.com/GSS-Cogs/idpd-api-poc/blob/main/src/schemas.py). These are used to (a) confirm data loaded in conforms to the data model and (b) responses from the api also confirm to the data model._

- [x] Datasets, Dataset
- [x] Editions, Edition
- [ ] Versions, Version
- [ ] Publishers, Publisher
- [ ] Topics

## Initial Data

_We have an (expanding) selection of jsonlds that are used to (a) power the stubbed initial data store and (b) are converted to pure RDF to populate the graph database._

- [x] Load datasets jsonld, enforce with schema.
- [x] Load _a_ editions jsonld, enforce with schema.
- [ ] Load _all_ sample editions in directory, not just one
- [ ] Load _all_ sample versions in directory, not just one. Apply versions schema during load.
- [ ] Enforce schemas during data load: publishers, topics.

## Sundry

_Small things that have gotten missed or "TODO:" in the opening stages that we need to clean up._

- [ ] Tests for /version and /versions endpoints. These are currently just the csv tests. We need the "application/json+ld" tests as well, as per editions and datasets.

---
Note: The above gets us an api with a stubbed store and test coverage. It will 100% look and work like the real deal but just has a bunch of hard coded files under the hood. This will do for getting going, but will become very unwieldy over time (managing lots of complex hard coded files) and doesn't _really_ meet the goal of proving we can do this with a graph.

TLDR: good enough to get going and a workable fallback for any endpoints we don't build out in full in the timescale, but not the real goal.

---

## Oxigraph Store

_This is moving beyond using the stub and implementing the OxigraphMetadataStore methods so that we can power the api directly using a graph database._

- [ ] (re) implement `OxigraphMetadataStore.get_dataset({id})` methods to power the `/datasets/{id{}` endpoint with data from the graph. Write tests.
- [ ] implement `OxigraphMetadataStore.get_datasets()` methods to power the `/datasets/` endpoint with data from the graph. Write tests.
- [ ] implement `OxigraphMetadataStore.get_edition({id})` methods to power the `/datasets/{id}/edition/{id}` endpoint with data from the graph. Write tests.
- [ ] implement `OxigraphMetadataStore.get_editions()` methods to power the `/datasets/{id}/editions` endpoint with data from the graph. Write tests.
- [ ] implement `OxigraphMetadataStore.get_version({id})` methods to power the `/datasets/{id}/editions/{id}/versions/{id}` endpoint with data from the graph. Write tests.
- [ ] implement `OxigraphMetadataStore.get_versions()` methods to power the `/datasets/{id}/editions/{id}/versions` endpoint with data from the graph. Write tests.
- [ ] implement `OxigraphMetadataStore.get_publisher({id})` methods to power the `/publisher/{id}/` endpoint with data from the graph. Write tests.
- [ ] implement `OxigraphMetadataStore.get_publishers()` methods to power the `/publisher` endpoint with data from the graph. Write tests.
- [ ] implement `OxigraphMetadataStore.get_topic({id})` methods to power the `/topics/{id}/`endpoint with data from the graph. Write tests.
- [ ] implement `OxigraphMetadataStore.get_topics()` methods to power the `/topics` endpoint with data from the graph. Write tests.

## CsvStore

_For our purposes here, we're assuming we'll store the csv tidy data files in GCP storage in a public bucket and the api will get them on request._

- [ ] Implement the api getting the csv files out of google cloud storage.

## Nice to have

_These are things we'll want and that add value but we can live without for the intial demo(s)._

- [ ] Structured logging inital implementation.
- [ ] Add structured logging statements to the api codebase.
- [ ] Configured the web version of the backend to write structured logs to GCP (the idea here is logs can then be searched through).
- [ ] Self documenting apis.

## Decisions

_Just for visibility._

- It's been agreed we can use a _flat_ graph for November rather than named graphs. Named graphs are (very likely) important for performance at scale but add compexity, given the relatively small data footprint for November it's been descoped (for now).
- Structures at time of writing are in progress. These will be iterated in parallel to developing the api codebase.
